File: Python3/0050. Pow(x, n).py
class Solution:
    def myPow(self, x: float, n: int) -> float:
        if n<0:
            return 1/self.myPow(x,-n)
        
        if n==0:
            return 1
        
        if n==1:
            return x
        
        half = self.myPow(x,n//2)
        res = self.myPow(x,n%2)

        return half*half*res

        # if n < 0:
        #     return 1/self.myPow(x,-n)
        
        # if n == 0:
        #     return 1
        
        # res = 1
        # while(n):
        #     power = 1
        #     t = x
        #     while(power<=n):
        #         n -= power
        #         res *= t
        #         t *= t
        #         power <<= 1
        
        # return res

File: C++/0050. Pow(x, n).cpp
class Solution {
public:
    double myPow(double x, long long n) {
        if(n==0){
            return 1.0;
        }
        else if(n<0){
            return 1.0/myPow(x,-n);
        }
        else{
            double y = myPow(x,n/2);
            return n%2==1?x*y*y:y*y; 
        }
    }
};
